frame/mosaic/src/lib.rs

```diff
@@ -0,0 +1,676 @@
+				asset_id,
+				|maybe_tx| match *maybe_tx {
+					Some((balance, lock_period)) => {
+						ensure!(amount <= balance, Error::<T>::AmountMismatch);
+						T::Assets::burn_from(
+							asset_id,
+							&Self::sub_account_id(SubAccount::outgoing(from.clone())),
+							amount,
+						)?;
+
+						// No remaing funds need to be transferred for this asset, so we can delete
+						// the storage item.
+						if amount == balance {
+							*maybe_tx = None
+						} else {
+							let new_balance =
+								balance.checked_sub(&amount).ok_or(Error::<T>::AmountMismatch)?;
+							*maybe_tx = Some((new_balance, lock_period));
+						}
+
+						Self::deposit_event(Event::<T>::TransferAccepted {
+							from,
+							asset_id,
+							amount,
+						});
+						Ok(())
+					},
+					None => Err(Error::<T>::NoOutgoingTx.into()),
+				},
+			)?;
+			Ok(().into())
+		}
+
+		/// Claims user funds from the `OutgoingTransactions`, in case that the relayer has not
+		/// picked them up.
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn claim_stale_to(
+			origin: OriginFor<T>,
+			asset_id: AssetIdOf<T>,
+			to: AccountIdOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let caller = ensure_signed(origin)?;
+
+			let now = <frame_system::Pallet<T>>::block_number();
+
+			OutgoingTransactions::<T>::try_mutate_exists(
+				caller.clone(),
+				asset_id,
+				|prev| -> Result<(), DispatchError> {
+					let amount = match *prev {
+						Some((balance, lock_time)) if lock_time < now => {
+							T::Assets::transfer(
+								asset_id,
+								&Self::sub_account_id(SubAccount::outgoing(caller.clone())),
+								&to,
+								balance,
+								false,
+							)?;
+							balance
+						},
+						_ => return Err(Error::<T>::NoStaleTransactions.into()),
+					};
+
+					*prev = None;
+					Self::deposit_event(Event::<T>::StaleTxClaimed { to, by: caller, amount });
+					Ok(())
+				},
+			)?;
+			Ok(().into())
+		}
+
+		/// Mints new tokens into the pallet's wallet, ready for the user to be picked up after
+		/// `lock_time` blocks have expired.
+		#[pallet::weight(10_000)]
+		pub fn timelocked_mint(
+			origin: OriginFor<T>,
+			asset_id: AssetIdOf<T>,
+			to: AccountIdOf<T>,
+			amount: BalanceOf<T>,
+			lock_time: BlockNumberOf<T>,
+			id: Id,
+		) -> DispatchResultWithPostInfo {
+			let (_caller, current_block) = ensure_relayer::<T>(origin)?;
+
+			AssetsInfo::<T>::try_mutate_exists::<_, _, DispatchError, _>(asset_id, |info| {
+				let AssetInfo { last_mint_block, penalty, budget, penalty_decayer } =
+					info.take().ok_or(Error::<T>::UnsupportedAsset)?;
+
+				let new_penalty = penalty_decayer
+					.checked_decay(penalty, last_mint_block, current_block)
+					.unwrap_or_else(Zero::zero);
+
+				let penalised_budget = budget.saturating_sub(new_penalty);
+
+				// Check if the relayer has a sufficient budget to mint the requested amount.
+				ensure!(amount <= penalised_budget, Error::<T>::InsufficientBudget);
+
+				T::Assets::mint_into(
```

Comment:
same question, are we sure relayer prevents double mint?

Reply:
this is a relayer concern